src/python/torchdistx/deferred_init.py

```diff
@@ -28,6 +28,12 @@ def deferred_init(module_fn: Callable[..., Module], *args, **kwargs) -> Module:
             ``Module`` instance.
         args, kwargs:
             The positional and keyword arguments to be passed to ``module_fn``.
+
+    .. Warning::
+        The operations performed on the parameters and buffers of a module will
+        only be recorded while inside ``deferred_init()``. Avoid making changes
+        to a module after its returned from ``deferred_init()``; otherwise it
+        cannot be correctly materialized.
```

Comment:
I tried the following in FSDP, not sure if it is supported, but should it work? 

```
class Holder(nn.Module):
    self.model = Transformer(...) # on CPU
    self.model = self.model.cuda()
```

holder = deferred_init(Holder, **model_args)

I guess this should work, but is sort of redundant as we can initialize directly on GPU instead of the copy?

Reply:
Yes, that would work, in fact you don't even need to have a `Holder` type.

```
def build_my_model(**kwargs):
    model = Transformer(**kwargs)
    return mode.cuda()

model = deferred_init(build_my_model, **model_args)
```

is cleaner.